### ISSUE TITLE
Type checking and return type inference

### ISSUE THREAD

**Issue by apurvaraman** [dcodeIO/AssemblyScript#69]
Are there any plans to include inferring the return type (rather than requiring it to be declared and assuming void otherwise) and checking whether the types of the parameters to a function are correct?

**Comment by dcodeIO** [comment]
With implicit type conversions in place, detecting a function's actual return type from its body could be error prone. The type system is much like that of C. My experience so far is that requiring explicit type definitions everywhere is the only coherent implementation choice we have.

> and checking whether the types of the parameters to a function are correct?

Of course, do you have a specific issue in this regard?

**Comment by apurvaraman** [comment]
Typescript has a type resolver that can evaluate the correct return type given a node in the AST. Typescript shouldn't be coercing the types, rather, it keeps track of the possible types that could be returned- in the case that wasm does not support returning a certain combination of types, an error could be thrown. 

For checking the parameters, the issue was passing the parameters from JS to wasm via an exported function, but the checking there would have to be done by the interface between JS and wasm.

**Comment by dcodeIO** [comment]
> Typescript has a type resolver that can evaluate the correct return type given a node in the AST.

Yeah, but it assumes JS types and doesn't work with `int`, `float` etc. (which are just aliases of `number`), effectively rendering it useless for our purpose.

> For checking the parameters, the issue was passing the parameters from JS to wasm via an exported function

The loader provides some utility to work with these, but doesn't provide any wrappers in its current state. Hence, you can't, for example, provide a `string` value from the JS side but have to allocate it in memory explicitly (here: `module.string.create`) before providing a pointer from JS to wasm.
